docs/building-modules/testing.md

```diff
@@ -0,0 +1,63 @@
+<!--
+order: 15
+-->
+
+# Module & App Testing
+
+The Cosmos SDK contains different types of [tests](https://martinfowler.com/articles/practical-test-pyramid.html).
+These tests have different goals and are used at different stages of the development cycle.
+We advice, as a general rule, to use tests at all stages of the development cycle.
+It is adviced, as a chain developer, to test your application and modules in a similar way than the SDK.
+
+The rationale behind testing can be found in [ADR-59](https://docs.cosmos.network/main/architecture/adr-059-test-scopes.html).
+
+## Unit Tests
+
+Unit tests are the lowest test category of the [test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html).
+All packages and modules should have unit test coverage. Modules should have their dependencies mocked: this means mocking keepers for example.
+
++++ https://github.com/cosmos/cosmos-sdk/blob/a92c291880eb6240b7221173282fee0c5f2adb05/x/gov/keeper/keeper_test.go#L21-L35
+
+The SDK uses `mockgen` to generate mocks for keepers:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/dd556936b23d7443cb7fb1da394c35117efa9da7/scripts/mockgen.sh#L12-L29
+
+## Integration Tests
```

Comment:
we should break out each section into a short tutorial/walkthrough on how to write these or link to something we have. This doesn't say much on how to do it.